### PR TITLE
PBF To Atlas Sub Commmand

### PR DESCRIPTION
**Pull request by Bentleysb**
### Description:
This adds a simple command to convert an OSM PBF to an Atlas file (as per https://github.com/osmlab/atlas/issues/88). It requires only the path to a PBF file, and a path for an output file. It has a number of optional parameters that can be passed to customize the Atlas generation. 

### Potential Impact:

No impacts for downstream libraries. 

### Unit Test Approach:
Unit test `testDefaultConversion` checks that that the command functions as intended when no optional parameters are passed.  
Unit test `testFiltersTextMapCountryCodesConversion` checks that all the filter parameters, country code parameter, and the country boundary map parameter with a text file work as expected.  
Unit test `testNoSectioningNoSlicingNoRelationsShapeMapConversion` checks that the way section, country slicing, load relation, and country boundary map (with a shape file) parameters work as expected.  
Unit test `testNoWaysConversion` checks that the load ways parameter works as expected. 

### Test Results:
Local tests were performed on multiple PBFs, using many combinations of default and optional parameters. All worked as expected, with the exception of using a large shape file as a country boundary map. This scenario caused the command to take a long time to load.  It is recommended to use text country boundary maps when possible. 

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/test/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    public void testDefaultConversion()
+    {
+        final File temp = File.temporaryFolder();
+
+        try
+        {
+            // Run PBFToAtlasSubCommand
+            final String[] args = { "pbf-to-atlas", String.format("-pbf=%s", PBF),
+                    String.format("-output=%s/%s", temp, ATLAS_NAME) };
+            new AtlasReader(args).runWithoutQuitting(args);
```
Comment (nachtm) [inline src/test/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommandTest.java]: Have you considered running the AtlasReaders in a  [@BeforeClass](https://junit.org/junit4/javadoc/4.12/org/junit/BeforeClass.html) method, then splitting up the tests into smaller chunks? The advantages to this are: 

- Purpose of each test will become clearer.
- Easier to distinguish between (for example) a way-sectioning problem and a country map problem if they're in separate tests.

Downsides to the @BeforeClass approach include:

- Storing all four atlases in memory at the same time.
- Tests become a little less independent.
- Harder to track down the source of runtime errors in actual AtlasReader runs.

Either way works, just food for thought!

**Review by nachtm — APPROVED** [review]
Looks great! Just one structural question re: tests.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                .fromBoundaryMap(Collections.singletonMap("UNK", MultiPolygon.MAXIMUM));
+        if (countryMapOption.isPresent())
+        {
+            if (FilenameUtils.isExtension(countryMapOption.get().getName(), "txt"))
+            {
+                countryMap = CountryBoundaryMap.fromPlainText(countryMapOption.get());
+            }
+            else if (FilenameUtils.isExtension(countryMapOption.get().getName(), "shp"))
+            {
+                countryMap = CountryBoundaryMap.fromShapeFile(countryMapOption.get().getFile());
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: It looks like `countryMapOption.get()` is fetched multiple times. Could make use of a variable for this file.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+     * @param map
+     *            {@link CommandMap} containing the {@code COUNTRY_MAP_PARAMETER}
+     * @return {@link CountryBoundaryMap} loaded from a file or default
+     */
+    private CountryBoundaryMap getCountryBoundaryMap(final CommandMap map)
+    {
+        final Optional<File> countryMapOption = (Optional<File>) map
+                .getOption(COUNTRY_MAP_PARAMETER);
+        CountryBoundaryMap countryMap = CountryBoundaryMap
+                .fromBoundaryMap(Collections.singletonMap("UNK", MultiPolygon.MAXIMUM));
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: Could make use of constant for string `UNK`.

**Review by sayas01 — APPROVED** [review]
This looks amazing! Great work @Bentleysb !

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java (hunk 3 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: I think this is fine, as it is only used the once.

File: src/test/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommandTest.java (hunk 1 above)
Comment (Bentleysb) [reply src/test/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommandTest.java]: I had not considered this. I like the idea of having a clearer purpose to the tests, and I don't think memory is an issue here. The thing that makes apprehensive is runtime error tracking. Having the atlas generation inside of tests helps to show if the different switches are functioning. So, for now I think I will leave them as is. However I am very open to further discussion. There is definitely room for improvement here.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
+package org.openstreetmap.atlas.geography.atlas.command;
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: PBF -> OsmPbf to stick with the already used convention with OsmPbf classes in the pbf command

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        writer.println(
+                "-way-sectioning=boolean : whether to perform way sectioning; defaults to true");
+        writer.println(
+                "-way-section-filter=/path/to/json/way/section/filter : json filter to determine where to way section");
+    }
+
+    @Override
+    public int execute(final CommandMap map)
+    {
+        new OsmPbfLoader((File) map.get(INPUT_PARAMETER), getAtlasLoadingOption(map))
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: Use `this` for local functions

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * @author bbreithaupt
+ */
+public class PBFToAtlasSubCommand implements FlexibleSubCommand
+{
+    private static final String NAME = "pbf-to-atlas";
+    private static final String DESCRIPTION = "Converts a PBF to an Atlas file.";
+
+    // Required parameters
+    private static final Command.Switch<File> INPUT_PARAMETER = new Command.Switch<>("pbf",
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: If you import `Switch` above, you wont have to call `Command.` every time you want to access the class

**Review by danielduhh — CHANGES_REQUESTED** [review]
Just had some naming suggestions

File: src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java (hunk 3 above)
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/command/PBFToAtlasSubCommand.java]: Even if it's used once, try and stray away from inline strings/numbers.

**Review by savannahostrowski — APPROVED** [review]
One small comment - LGTM.

**Review by danielduhh — APPROVED** [review]

**Comment by danielduhh** [comment]
Nice work Bentley !

**Pull request closed without merging**